### ISSUE TITLE
NOTIFY received before call is answered cannot be delivered to the controlling app

### ISSUE THREAD

**Issue by CelsoGalo** [davehorton/drachtio-server#75]
Hello Dave, hope are doing good :-)

I have a problem with this scenario: My Drachtio server is receiving a NOTIFY while the call is still being advertised (I mean, it is in ringing state) and that NOTIFY contains the same Call-Id, To and From headers (including to-tag and from-tag) as indicated in the 180 RINGING.

The problem is that Drachtio does not allow at all this message to be advertised to the controlling NodeJS application. Please note that on my app I have handling code for in-dialog and out-of-dialog NOTIFYs, but none is called. The Drachtio server logs show the following:
```
SipDialogController::processRequestInsideDialog: NOTIFY irq 0x2891970
ClientController::route_request_inside_dialog - client managing dialog has disconnected: 682a3c6c-e210-1237-aea0-080027c3e781;uas
SipDialogController::addIncomingRequestTransaction - adding transactionId 4817825a-3368-41b5-a34e-39fd7fd9fd06 for irq:0x2891970
```
From the msg "client managing dialog has disconnected" I would infer that my NodeJS app had disconnected during the handling, but actually by looking the network traces (attached) we can see that this did not happen.

Dave, maybe we have here something in the Drachtio design that could be enhanced. This NOTIFY appears really to be marked to be handled inside the dialog, but as the call is not answered, actually there is no dialog object on my NodeJS application so far... On my code the dlg object is only available after the call is answered, which is not the case here.
```
    dlgProm.then(dlg => {
        logger.info(`Dialog established, remote uri is ${dlg.remote.uri}`);
        this.dlg = dlg;

        // Other handling here...

        dlg.on('notify', (/*req, res*/) => {
            logger.info('In-dialog NOTIFY has been received');
        });
    })
```
My suggestion would be: In case the call is not answered, can't Drachtio server deliver the NOTIFY as an out-of-dialog request?

This is a critical project and the team need a response from me, thank you so much for helping!

**Comment by CelsoGalo** [comment]
Here you find the Drachtio server logs and a network trace.
[drachtio-latest.log](https://github.com/davehorton/drachtio-server/files/3117737/drachtio-latest.log)

**Comment by davehorton** [comment]
There are no actual SIP messages in that log, which is strange.  Any idea why that is?  Were they removed?  I really need to see the SIP traffic to troubleshoot this

**Comment by CelsoGalo** [comment]
Hey Dave, please look the wireshark trace inside the zip file. The NOTIFY is the #41 package in that trace.

**Comment by davehorton** [comment]
OK, it took a while for that file to upload so I didnt see it at first.

So the drachtio server is at 192.168.7.15 in this trace, and is receiving the INVITE and then receiving the NOTIFY, correct?

**Comment by davehorton** [comment]
I think you are probably right, since the INVITE has not been answered with 200 OK or a reliable provisional response yet, the NOTIFY should probably be delivered as an out of dialog request.

**Comment by CelsoGalo** [comment]
Correction, Drachtio server is listening at 192.168.7.15:5060 and receiving the INVITE + NOTIFY without the call being answered.

**Comment by davehorton** [comment]
ok I may have a fix for this later tonight.  Stand by...

**Comment by CelsoGalo** [comment]
Thank you so much! BTW, just a comment about your comment:

_I think you are probably right, since the INVITE has not been answered with 200 OK or a reliable provisional response yet, the NOTIFY should probably be delivered as an out of dialog request._

Drachtio server has sent already a 180 RINGING before receiving the NOTIFY, but the dialog object is anyway not available, right?

**Comment by davehorton** [comment]
Can you possibly test a patch for me on this?  I've checked in a fix on a new branch 'notify-during-call-setup'.  If you are running using a native build, would you be able to build and test your scenario against that?  If you are using docker there should (shortly) be a docker image at `drachtio/dachtio-server:notify-during-call-setup` that has this patch.

The expected behavior is that you will still see the misleading message about client has disconnected when the NOTIFY comes in, but then it should be forwarded to your app as an out of dialog request (e.g. you will need an `srf.notify()` handler.

I haven't been able to test this scenario fully myself today due to some competing priorities, so getting your results will be helpful

**Comment by CelsoGalo** [comment]
Thanks a lot, Dave! It is working fine!

I am adding logs and trace for you.
[successful logs and traces.zip](https://github.com/davehorton/drachtio-server/files/3121750/successful.logs.and.traces.zip)

BTW, I did a quick regression using my Unit tests - they were all fine also.

I keep the ticket open for you to close once merged, right? Once more, thank you so much!!

**Comment by davehorton** [comment]
great and many thanks for the logs.  That will help.  I will run some further tests this weekend and look to merge into the develop branch over the weekend.  Leaving open till then

**Comment by davehorton** [comment]
just out of interest, how is it that the SIP traces do not appear in the drachtio log file?  Are you scrubbing them from the log before posting (that's fine, I just want to make sure there is no issue because I have it set up to pretty much always log the SIP traces since I rely on them so much)

**Comment by CelsoGalo** [comment]
Oh, sorry Dave... Indeed I was filtering the original file based in the timestamp, so this was excluding the SIP msgs dump. I will be more careful next time :-), but here is the successful logs again without that bad filter.
[drachtio-with-sip.log](https://github.com/davehorton/drachtio-server/files/3122032/drachtio-with-sip.log)

**Comment by davehorton** [comment]
merged into develop branch as v0.8.1-rc2
